generic/tests/ping.py

```diff
@@ -58,76 +64,141 @@ def _get_loss_ratio(output):
+            if output.find(result_str) == -1:
+                test.fail(("Ping test with exceeded package failed,"
+                           " expected string: '%s' not found" %
+                           result_str))
+            # ping with flood
+            error_context.context("Flood ping test from host",
+                                  logging.info)
+            utils_test.ping(ip, None, flood=True, output_func=None,
+                            interface=interface, timeout=flood_minutes * 60)
+
+            error_context.context("Ping test after flood ping,"
+                                  " Check if the network is still alive",
+                                  logging.info)
+            status, output = utils_test.ping(ip, counts, interface=interface,
+                                             timeout=float(counts) * 1.5)
             _get_loss_ratio(output)
 
             if status != 0:
-                raise error.TestFail("Ping failed, status: %s,"
-                                     " output: %s" % (status, output))
+                test.fail("Ping returns non-zero value %s" % output)
 
-        error.context("Flood ping test", logging.info)
-        utils_test.ping(ip, None, flood=True, output_func=None,
-                        interface=interface, timeout=flood_minutes * 60)
+    if ping_ext_host:
+        error_context.context("Ping test from guest side,"
+                              " dest: '%s'" % ext_host, logging.info)
+        pkt_sizes = packet_sizes
+
+        is_windows = False
+        if params.get("os_type") == "windows":
+            # Windows doesn't support ping with packet larger than '65500'
+            pkt_sizes = [p for p in packet_sizes if int(p) < 65500]
+            # Add a packet size just equal '65500' for windows
+            pkt_sizes.append(65500)
+            is_windows = True
+
+        for size in pkt_sizes:
+            if is_windows:
+                logging.info("Ping with packet size: %s" % size)
+                status, output = utils_test.ping(ext_host, 30, packetsize=size,
+                                                 session=session, timeout=60)
+                _get_loss_ratio(output)
+                if status != 0:
+                    test.fail(("Ping external host failed,"
+                               " status: %s, output: %s" %
+                               (status, output)))
+            else:
+                for interval_t in interval_times:
+                    logging.info("Ping with packet size: %s and interval: %s" %
+                                 (size, interval_t))
+                    status, output = utils_test.ping(ext_host, 30,
+                                                     packetsize=size,
+                                                     interval=interval_t,
+                                                     session=session,
+                                                     timeout=60)
```

Comment:
I think this part could be put into one function. 
You can use params to control how it works.